ee/tabby-webserver/src/service/auth.rs

```diff
@@ -173,19 +177,57 @@ impl AuthenticationService for DbConn {
             return Err(TokenAuthError::InvalidPassword);
         }
 
+        let utc_ts = chrono::Utc::now().timestamp();
+        let refresh_token = match self.get_user_unexpired_token(user.id, utc_ts).await? {
+            Some(refresh_token) => refresh_token.token,
+            None => {
+                let (token, expires_in) = generate_refresh_token(utc_ts);
+                self.create_refresh_token(user.id, &token, expires_in)
+                    .await?;
+                token
+            }
+        };
+
         let Ok(access_token) = generate_jwt(Claims::new(UserInfo::new(
             user.email.clone(),
             user.is_admin,
         ))) else {
             return Err(TokenAuthError::Unknown);
         };
 
-        let resp = TokenAuthResponse::new(access_token, "".to_string());
+        let resp = TokenAuthResponse::new(access_token, refresh_token);
         Ok(resp)
     }
 
-    async fn refresh_token(&self, _refresh_token: String) -> Result<RefreshTokenResponse> {
-        Ok(RefreshTokenResponse::default())
+    async fn refresh_token(
+        &self,
+        token: String,
+    ) -> std::result::Result<RefreshTokenResponse, RefreshTokenError> {
+        let Some(refresh_token) = self.get_refresh_token(&token).await? else {
+            return Err(RefreshTokenError::InvalidRefreshToken);
+        };
+        if refresh_token.is_expired() {
+            return Err(RefreshTokenError::ExpiredRefreshToken);
+        }
+        let Some(user) = self.get_user(refresh_token.user_id).await? else {
+            return Err(RefreshTokenError::UserNotFound);
+        };
+
+        // refresh token is verified, generate new access token based on user info
+        let Ok(access_token) = generate_jwt(Claims::new(UserInfo::new(
+            user.email.clone(),
+            user.is_admin,
+        ))) else {
+            return Err(RefreshTokenError::Unknown);
+        };
+
+        let resp = RefreshTokenResponse::new(
+            access_token,
+            refresh_token.token,
```

Comment:
is it a better practice to return a updated refresh token?

Reply:
That's a good question, I had the same thought.

With some google, I read a point of view that this could cause security issue since there's no way to prevent user from abusing this API, to generate huge amounts of refresh tokens.

By always returning new token, I see a few possible issues:
- user will never be logged out  (unless we revoke all of them)
- db rows grow unexpectedly
- when login, not sure which refresh token to return.
- when revoke, might cause revoke partially done
  - say user login A device with refresh token A, B device with refresh token B. when revoke A, B device still have access

Technically, it's definitely ok.

Reply:
> user will never be logged out (unless we revoke all of them)
Why? A user will be logged out as long as his current refresh token expires (since he need to re-login by using token auth)

> db rows grow unexpectedly
agree

> when login, not sure which refresh token to return.
should be a new token.

> when revoke, might cause revoke partially done
> say user login A device with refresh token A, B device with refresh token B. when revoke A, B device still have access
This behavior could easily adjust given the requests:
1. If we wanna device B being logged out, we should just revoke all refresh token belong to this user.
2. If we wanna device B stayed logged in, we'll just leave it as is.

Isn't this part of the reason we make refresh token a separate table to make each of it can be revoke individually?

Reply:
> > user will never be logged out (unless we revoke all of them)
> 
> Why? A user will be logged out as long as his current refresh token expires (since he need to re-login by using token auth)

I mean a scenario like this: say user login, got refresh token A, user do refresh with A, got new access token and new refresh token B, B expires later than A, so user can continue to use B to get access token and another refresh token C, something like that, is it possible?

> > db rows grow unexpectedly
> 
> agree, this is problematic, and seems the only way is to apply certain rate limiting.
> 
> > when login, not sure which refresh token to return.
> 
> should be a new token.
> 
> > when revoke, might cause revoke partially done
> > say user login A device with refresh token A, B device with refresh token B. when revoke A, B device still have access
> 
> This behavior could easily adjust given the requests:
> 
> 1. If we wanna device B being logged out, we should just revoke all refresh token belong to this user.
> 2. If we wanna device B stayed logged in, we'll just leave it as is.
> 
> Isn't this part of the reason we make refresh token a separate table to make each of it can be revoke individually?

hmm...ya, I remember it is. How about return new refresh token for each login (i.e. `token_auth`), but not for `refresh_token`?

We put rate limit on login request, so user won't abuse it, make sense?

Reply:
> I mean a scenario like this: say user login, got refresh token A, user do refresh with A, got new access token and new refresh token B, B expires later than A, so user can continue to use B to get access token and another refresh token C, something like that, is it possible?

I think this is the average practice - and it's easy to adjust the behavior (because we're updating the same row, we control the new expire time at refresh time anyway

> How about return new refresh token for each login (i.e. token_auth), but not for refresh_token?

for refresh token - we're not creating new row in refresh token, we only update the value of `token.

> We put rate limit on login request, so user won't abuse it, make sense?

No need to implement this in this PR